proofread: cache_management.md (#2478)
juicedata/juicefs: docs/en/guide/cache_management.md

JuiceFS provides various caching mechanisms including metadata caching, data read/write caching, etc.

## Data consistency

JuiceFS provides a "close-to-open" consistency guarantee, which means that when two or more clients read and write the same file at the same time, the changes made by client A may not be immediately visible to client B. However, once the file is closed by client A, any client re-opened it afterwards is guaranteed to see the latest data, no matter it is on the same node with A or not.

"Close-to-open" is the minimum consistency guarantee provided by JuiceFS, and in some cases it may not be necessary to reopen the file to access the latest written data. For example, multiple applications use the same JuiceFS client to access the same file (where file changes are immediately visible), or to view the latest data on different nodes with the `tail -f` command.

## Cache read mechanism

> JuiceFS provides a multi-level cache to improve the performance of accessing frequently used files. Read requests will try the kernel page cache, the pre-read buffer of the JuiceFS process, and the local disk cache in turn. If the data requesting is not found in any level of the cache, it will be read from the object storage, and also be written into every level of the cache asynchronously to improve the performance of the next access.

Reviewer: ```suggestion
JuiceFS provides a multi-level cache to improve the performance of frequently accessed data. Read requests will try the kernel page cache, the pre-read buffer of the JuiceFS process, and the local disk cache in turn. If the data requested is not found in any level of the cache, it will be read from the object storage, and also be written into every level of the cache asynchronously to improve the performance of the next access. 
```

![](../images/juicefs-cache.png)

## Metadata cache

JuiceFS supports caching metadata both in kernel and in client memory (i.e. JuiceFS processes) to improve metadata access performance.

> ### Kernel metadata cache

Reviewer: ```suggestion
### Metadata cache in Kernel
```
原来的翻译是不是更好些？ :P

### In-memory client metadata cache

:::note
This feature requires JuiceFS >= 0.15.2
:::

When a JuiceFS client `open()` a file, the attributes of the file are automatically cached in client memory. If the [`--open-cache`](../reference/command_reference.md#juicefs-mount) option is set to a value greater than 0 when mounting the file system, subsequent `getattr()` and `open()` operations will return the result from the in-memory cache immediately, as long as the cache has not expired.

When dong `read()` on a file, the chunk and slice information of the file is automatically cached in the client memory. Reading the chunk again during the cache lifetime will return the slice information from the in-memory cache immediately.

:::tip
You can check ["How JuiceFS Stores Files"](../reference/how_juicefs_store_files.md) to know what chunk and slice are.
:::

> By default, for a file whose metadata has been cached in memory and not being accessed by any process for more than 1 hour, all its metadata cache will be automatically deleted.

Reviewer: ```suggestion
By default, if there is a file whose metadata has been cached in memory and not accessed by any process for more than 1 hour, all its metadata cache will be automatically deleted.
```

Author: 这里用 `being` 我想表达的是“正在被访问”，大概是 `and (which is) not being accessed` 这种形式，你看 ok 不？

Author: > 

这里用 being 我想表达的是“正在被访问”，大概是 and (which is) not being accessed 这种形式，你看 ok 不？

## Data cache

> JuiceFS also provides kinds of data cache, including page cache in the kernel and local cache on the client host, to improve performance.

Reviewer: ```suggestion
JuiceFS also provides various kinds of data cache, including page cache in the kernel and local cache on the client host, to improve performance.
```

### Kernel data cache

:::note
This feature requires JuiceFS >= 0.15.2
:::

For a file that has already been read, the kernel automatically caches its content. If this file is not updated (i.e. its `mtime` doesn't change) in the further accessings, it will be read directly from the kernel cache to obtain better performance.

Thanks to the kernel cache, repeated reads of the same file in JuiceFS can be extremely fast, with latencies as low as a few microseconds and throughputs up to several GiBs per second.

> JuiceFS clients currently do not have kernel write caching enabled by default. Starting with [Linux kernel 3.15](https://github.com/torvalds/linux/commit/4d99ff8f12e), FUSE supports ["writeback-cache mode"]( https://www.kernel.org/doc/Documentation/filesystems/fuse-io.txt), which means that the `write()` system call can be done very quickly. You can set the [`-o writeback_cache`](../reference/fuse_mount_options.md#writeback_cache) option at [mounting file system](../reference/command_reference.md#juicefs-mount) to enable writeback-cache mode. It is recommended to enable this mount option when very small data (e.g. around 100 bytes) needs to be written frequently.

Reviewer: ```suggestion
JuiceFS clients currently do not support kernel write caching enabled by default. Starting with [Linux kernel 3.15](https://github.com/torvalds/linux/commit/4d99ff8f12e), FUSE supports ["writeback-cache mode"]( https://www.kernel.org/doc/Documentation/filesystems/fuse-io.txt), which means that the `write()` system call can be done very quickly. You can set the [`-o writeback_cache`](../reference/fuse_mount_options.md#writeback_cache) option at [mounting file system](../reference/command_reference.md#juicefs-mount) to enable writeback-cache mode. It is recommended to enable this mount option when very small data (e.g. around 100 bytes) needs to be written frequently.
```

Author: 这里后面有 enabled 了，所以前面用 have 比较合适吧，如果是 support 的话要删掉后面的 enabled？

### Client read data cache

The JuiceFS client automatically prefetches data into the cache based on the read pattern, thus improving sequential read performance. By default, it prefetches 1 block concurrently and caches it locally when reading. The local cache can be set up on any local file system based on HDD, SSD or memory.

:::tip Hint
You can check ["How JuiceFS Stores Files"](../reference/how_juicefs_store_files.md) to learn what a block is.
:::

The local cache can be adjusted at [mounting file system](../reference/command_reference.md#juicefs-mount) with the following options.

```
--cache-dir value         directory paths of local cache, use colon to separate multiple paths (default: "$HOME/.juicefs/cache" or "/var/jfsCache")
--cache-size value        size of cached objects in MiB (default: 102400)
--free-space-ratio value  minimum free space (ratio) (default: 0.1)
--cache-partial-only      cache only random/small read (default: false)
```

> Specifically, there are two ways to store the local cache of JuiceFS in memory -- setting `--cache-dir` to `memory` or `/dev/shm/<cache-dir>`. The difference between these two approaches is that, the former deletes the cache data after remounting the JuiceFS file system, while the latter retains it. There is not much difference in performance between them.

Reviewer: ```suggestion
Specifically, there are two ways to store the local cache of JuiceFS in memory, i.e., setting `--cache-dir` to `memory` or `/dev/shm/<cache-dir>`. The difference between these two approaches is that the former deletes the cache data after remounting the JuiceFS file system, while the latter retains the cache data. Still, there is not much difference in performance between them.
```

> The JuiceFS client writes data downloaded from the object store (including new uploaded data containing less than 1 block) to the cache directory as fast as possible, without compression or encryption. **Because JuiceFS generates unique names for all block objects written to the object store, and all block objects are not modified, there is no need to worry about the cached data invalidation when the file content is updated.**

Reviewer: ```suggestion
The JuiceFS client writes data downloaded from the object storage (including new uploaded data containing less than 1 block) to the cache directory as fast as possible, without compression or encryption. **Because JuiceFS generates unique names for all block objects written to the object storage, and all block objects are not modified, there is no need to worry about the invalidation of the cached data when the file content is updated.**
```

Data caching can effectively improve the performance of random reads. For applications like Elasticsearch, ClickHouse, etc. that require higher random read performance, it is recommended to point the cache path to a faster storage medium and allocate more cache space.

#### Cache life cycle

When the cache usage space reaches the upper limit (that is, the cache size is greater than or equal to `--cache-size`) or the disk is almost full (that is, the free disk space ratio is less than `--free-space-ratio`), the JuiceFS client will automatically clean the cache.

> JuiceFS allocates 100GiB cache space by default, but this does not mean that you have to have a disk with a capacity more than that. This value represents the maximum capacity that a JuiceFS client may use if the disk capacity allows. When the remaining space on the disk is less than 100GiB, ensure that the remaining space is not less than 10% by default.

Reviewer: ```suggestion
JuiceFS allocates 100GiB cache space by default, but this does not mean that you have to have a disk with a capacity more than that. This value represents the maximum capacity that a JuiceFS client may use if the disk capacity allows. When the remaining space on the disk is less than 100GiB, JuiceFS ensures that the remaining space is not less than 10% by default.
```

Author: 感觉改了下意思正好相反了😂。原意应该是说，juicefs 会保留至少 10% 的可用空间，而不是至少分配 10% 的缓存空间。

For example, if you set `--cache-dir` to a partition with a capacity of 50GiB, then if `--cache-size` is set to 100GiB, JuiceFS will always keep the cache capacity around 45GiB, which means that more than 10% of the partition is reserved.

When the cache is full, the JuiceFS client cleans the cache using LRU-like algorithm, i.e., it tries to clean the oldest and least-used data.

### Client write data cache

When writing data, the JuiceFS client caches the data in memory until it is uploaded to the object storage when a chunk is written or when the operation is forced by `close()` or `fsync()`. When `fsync()` or `close()` is called, the client waits for data to be written to the object storage and notifies the metadata service before returning, thus ensuring data integrity.

> In some cases where local storage is reliable and its write performance is significantly better than network writes (e.g. SSD disks), write performance can be improved by enabling asynchronous upload of data so that the `close()` operation does not wait for data to be written into the object storage, but returns as soon as the data is written to the local cache directory.

Reviewer: ```suggestion
When local storage is reliable and its write performance is significantly better than network writes (e.g. SSD disks), write performance can be improved by enabling asynchronous upload of data. In this way, the `close()` operation does not have to wait for data to be written into the object storage, but returns as soon as the data is written to the local cache directory.
```

When asynchronous upload is enabled, i.e. `--writeback` is specified when mounting the file system, do not delete the contents in `<cache-dir>/<UUID>/rawstaging` directory, as this will result in data loss.
:::

When the cache disk is almost full, it will pause writing data and change to uploading data directly to the object storage (i.e., the client write cache feature is turned off).

> When the asynchronous upload function is enabled, the reliability of the cache itself is directly related to the reliability of data writing, and should be used with caution for scenarios requiring high data reliability.

Reviewer: ```suggestion
When the asynchronous upload function is enabled, the reliability of the cache itself is directly related to the reliability of data writing. Thus, this function should be used with caution for scenarios requiring high data reliability.
```

- **macOS**: `$HOME/.juicefs/cache`
- **Windows**: `%USERPROFILE%\.juicefs\cache`

For Linux, note that the default cache path requires administrator privileges and that normal users need to be granted to use `sudo` to set it up, e.g.:

```shell
sudo juicefs mount redis://127.0.0.1:6379/1 /mnt/myjfs
```

> Alternatively, the `--cache-dir` option can be set to any storage path accessible to the current system when mounting the filesystem. For normal users who do not have permission to access the `/var` directory, the cache can be set in the user's `HOME` directory, e.g.:

Reviewer: ```suggestion
Alternatively, the `--cache-dir` option can be set to any storage path accessible by the current system when mounting the filesystem. For normal users who do not have permission to access the `/var` directory, the cache can be set in the user's `HOME` directory, e.g.:
```

Author: 这里应该用 to 更合适一些。`--cache-dir` 的值是 `any storage path accessible to the current system`

```shell
juicefs mount --cache-dir ~/jfscache redis://127.0.0.1:6379/1 /mnt/myjfs
```

:::tip
Setting up the cache on a faster SSD disk can improve performance.
:::

### RAM disk

> If you require higher file read performance, you can set up the cache into the RAM disk. For Linux systems, check the `tmpfs` file system with `df` command.

Reviewer: ```suggestion
If a higher file read performance is required, you can set up the cache into the RAM disk. For Linux systems, check the `tmpfs` file system with the `df` command.
```

> When multiple cache paths are set, the client will write data evenly to each cache path using hash policy.

Reviewer: ```suggestion
When multiple cache paths are set up, the client will write data evenly to each cache path using hash policy.
```

Author: 这里应该是 set 了。path 是参数，set 是指设置参数，set up 是搭建的意思

> JuiceFS currently estimates the size of cached objects by adding up the size of all cached objects and adding a fixed overhead (4KiB), which is not exactly the same as the value obtained by the `du` command.

Reviewer: ```suggestion
JuiceFS currently estimates the size of cached objects by adding up the size of all cached objects plus a fixed overhead (4KiB), which is not exactly the same as the value obtained by the `du` command.
```

> To prevent the cache disk from being written to full, the client will try to reduce the cache usage when the file system the cache directory is located on is running out of space.

Reviewer: ```suggestion
To prevent the cache disk from being overwritten, the client will try to reduce the cache usage when the file system on which the cache directory is located is running out of space.
```

Author: `overwrite` 这个词比较特殊，计算机领域里它的意思是「重写」或者「覆盖」，我们要不要想想还有没有更合适的？
